api/wft-complete-metadata.md

```diff
@@ -0,0 +1,126 @@
+# Workflow Task Completion Metadata
+
+We have a need to attach various bits of metadata to workflow task completions. There are a number
+of usecases in which the SDK would like to save some information for itself, or tell server about
+something, but this information does not influence replay or the reconstitution of workfow state in
+any way. Thus, it shouldn't go in a marker or some new kind of history event, because this would
+violate the (useful) idea that history events are primarily concerned with recreating workflow
+state.
+
+## Use cases
+
+### Reporting currently-being-awaited commands & facts about them
+
+Right now, besides a query, there's no simple way for a user to understand what their workflow is
+actively waiting on. Solving this would make it much easier for users to have insight into why
+their workflow is or isn't making progress, and could be nicely exposed in the UI.
+
+In particular, understanding what's going on while your workflow is currently running local
+activities and performing WFT heartbeats is very opaque. This would make that much more clear.
+
+Imagine being able to get information like this (and do useful things with it in the UI):
+
+```
+waiting_on: [
+    { command_type: "local_activity", activity_name: "la_func", id: 1, retries: 0, location: "my_file.py:23" },
+    { command_type: "local_activity", activity_name: "la_func", id: 2, retries: 5, location: "my_file.py:23" },
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
+```
+
+### SDK internal versioning
+
+We would like to be able to make SDK changes which affect how history is written, which means
+flagging our behavior based on what version of the SDK the history was produced with. Note: *not*
+relating to concepts of versioning a *user's* code, but rather the SDK itself.
+
+This internal version could easily slot into this metadata.
+
+Note: This could potentially go in `temporal.api.taskqueue.v1.VersionId` too, but that's more
+intended to relate to user code.
+
+### Dynamic config adjustments
```

Comment:
I don't see the use case for making dynamic config adjustments in the WFT completion event. There are already existing methods to modify the dynamic config. Is the idea here that you can make it more atomic somehow? If so, that's something users may end up relying on. So we'd need to maintain that contract on the server-side, which seems error-prone to me and difficult to future-proof.

Reply:
Maybe it wasn't clear, but these are all _possible_ use cases. The only one we know we need now is the SDK internal versioning ability, and probably the ability to track outstanding commands. Everything else is in "could be supported by this if we want" territory. Same applies for your subsequent comments, but I'll mention my thoughts on those too.

As mentioned here https://github.com/temporalio/proposals/pull/73#discussion_r1068742307 - yeah, maybe this one we'd avoid.